### PR TITLE
Signature Verification

### PR DESCRIPTION
**Pull request by sfyll**
Cairo function to handle signature verification. Right now we take the functional route, whereby we export a function which will run using a given public key at deploy time.

In the future, to be able to modify this public key when required, we can take an object-oriented approach, i.e. use smart contracts and "onlyOwner" type of construct.

One can find in /provider the code used to generate the signature. Please note that this is a hint we will try to follow [EIP-1193](https://eips.ethereum.org/EIPS/eip-1193#request-1) guidelines to reduce friction as much as possible for the user.

### CODE REVIEW

**Comment by lyronctk** [comment]
Agreed on the current functional route. I like this as a default since we haven't gauged customer comfort with onlyOwner functions

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+node_modules/
+target/
+.DS_Store
```
Comment (lyronctk) [inline .gitignore]: nit: newline

File: client/provider/provider.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { ethers } from 'ethers';
+
+// Provided private key
+const privateKey = '0x89ac8b3450f4f77fba1214fde42da15bc0d6b74cd1e77d8831528143ed8dc52c';
```
Comment (lyronctk) [inline client/provider/provider.ts]: could you add as an env variable

it's one of these bits where 1/1000 times you upload a real privkey. rather not take the unnecessary risk

File: client/provider/provider.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+// The message to sign
+const message = 'Hello, world!';
+
+// Hash the message (Ethereum's message hashing standard)
+const messageHash = ethers.hashMessage(message);
+
+const signature_data = wallet.signingKey.sign(messageHash);
+
+console.log("signature data: ", signature_data);
```
Comment (lyronctk) [inline client/provider/provider.ts]: we like to use tiered logging :)

e.g. the right side of my terminal at time 07:23 
(https://www.loom.com/share/a8d508ae3a3f4755a3374130da10aef1?sid=ca46ed9f-85f4-49ae-8058-7b3f21dc1a5d)

File: contract/src/signature_verifier.cairo [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+const public_key : felt252 =  0x30A333e4207129AD466a0B88b329651833605de4;
+
+fn verify_signature(commitment: u256, signature: Signature) {
+    let eth_address: EthAddress = public_key.try_into().unwrap();
+    return is_eth_signature_valid(commitment, signature, eth_address).unwrap();
+}
+
+#[test]
+fn test_verify_signature() {
```
Comment (lyronctk) [inline contract/src/signature_verifier.cairo:18]: excellent test coverage! our testing policy is usually extremely thorough integration tests + unit tests when necessary

i'd say this is on the edge of whether or not to write a unit test due to logic mostly being contained in imported functions, but i'll never complain about additional testing :)

File: contract/src/signature_verifier.cairo [hunk 5]
```diff
@@ -0,0 +1,38 @@
+use starknet::eth_signature::{is_eth_signature_valid};
+use starknet::EthAddress;
+use starknet::secp256_trait::Signature;
+
+
+//| Account address |  0x30A333e4207129AD466a0B88b329651833605de4
+//| Private key     |  0x89ac8b3450f4f77fba1214fde42da15bc0d6b74cd1e77d8831528143ed8dc52c
+//| public key      |  0x02d8834cfd7957d3cdf4ddd4ff038e6f8c06f31170d486cb52b737c0d41b0273e0
+
+const public_key : felt252 =  0x30A333e4207129AD466a0B88b329651833605de4;
```
Comment (lyronctk) [inline contract/src/signature_verifier.cairo:9]: could we set this as an env var as well

File: client/provider/provider.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+    const recoveredAddress = ethers.recoverAddress(messageHash, signature);
+
+    const isValid = recoveredAddress.toLowerCase() === wallet.address.toLowerCase();
+    console.log('Is the signature valid?', isValid);
+}).catch(error => {
+    console.error('Error signing the message:', error);
+});
+
+
```
Comment (lyronctk) [inline client/provider/provider.ts]: nit: extra whitespace

File: client/provider/provider.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    console.log('Signature:', signature);
+
+    const recoveredAddress = ethers.recoverAddress(messageHash, signature);
+
+    const isValid = recoveredAddress.toLowerCase() === wallet.address.toLowerCase();
+    console.log('Is the signature valid?', isValid);
+}).catch(error => {
+    console.error('Error signing the message:', error);
+});
+
```
Comment (lyronctk) [inline client/provider/provider.ts]: a great way to wrap up this PR would be to broadcast a function call to the cairo contract from typescript 

then we can port that logic from the provider to the client when do the changes there

perhaps it's even best to put together a dummy typescript client so you can run through an entire demo of what the provider<>client<>contract interaction is prior to touching RYO's official client

File: contract/src/signature_verifier.cairo [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+fn verify_signature(commitment: u256, signature: Signature) {
```
Comment (lyronctk) [inline contract/src/signature_verifier.cairo:12]: our doc standard is to have natural language paragraphs (generous term, anything between 1-3 sentences is sufficient) above every function call 

then comments within the functions are good whenever helpful, but not required

File: client/provider/provider.ts (hunk 3 above)
Comment (lyronctk) [reply client/provider/provider.ts]: if this is temporary scaffolding so you can test the cairo contract, you can wrap inside a function called temporary

File: client/provider/provider.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+const message = 'Hello, world!';
+
+// Hash the message (Ethereum's message hashing standard)
+const messageHash = ethers.hashMessage(message);
+
+const signature_data = wallet.signingKey.sign(messageHash);
+
+console.log("signature data: ", signature_data);
+
+wallet.signMessage((message)).then(signature => {
```
Comment (lyronctk) [inline client/provider/provider.ts]: please feel free to take advantage of as much of the code we already have written as possible! 

https://github.com/SeismicSystems/pilot-tomo/blob/main/src/utils/signature.ts
^ just added you as a reader of this private repository

File: .gitignore (hunk 1 above)
Comment (sfyll) [reply .gitignore]: https://github.com/sfyll/pilot-ryo/pull/1/commits/2a86778dc5131843ac1f2d74e10cf069cde0a238

File: client/provider/provider.ts (hunk 3 above)
Comment (sfyll) [reply client/provider/provider.ts]: deleted, was purely for testing the sc.

Will use what you guys cooked when implementing actual signature.

File: client/provider/provider.ts (hunk 7 above)
Comment (sfyll) [reply client/provider/provider.ts]: agreed, goal is to have a working signature req<>response pattern via dummy typescript before fiddling into the actual front-end!

File: client/provider/provider.ts (hunk 6 above)
Comment (sfyll) [reply client/provider/provider.ts]: delete file, cf above

File: contract/src/signature_verifier.cairo (hunk 5 above)
Comment (sfyll) [reply contract/src/signature_verifier.cairo:9]: https://github.com/sfyll/pilot-ryo/pull/1/commits/893bfb7d90daee311d688c92571fcb8107952a86

**Review by lyronctk — APPROVED** [review]
LGTM! great first PR man

approving with the last nit: we typically like commenting functions using multi-line comments 

so 
```
// Verify Seismic signature over a commitment.
```
becomes 
```
/*
 * Verify Seismic signature over a commitment.
 */
```

**Pull request merged**